Add torch._C._BUILD_NAMEDTENSOR()

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23625 Implement tensor.align_to(names), torch.align_tensors(*tensors)
* #23624 Add name propagation for at::alias, add tensor.set_names
* #23316 Add names to repr for named tensors
* **#23623 Add torch._C._BUILD_NAMEDTENSOR()**

This is a quick, not-user-facing check for if pytorch was built with BUILD_NAMEDTENSOR=1.

Test Plan:
- run tests [namedtensor ci]

gh-metadata: pytorch pytorch 23623 gh/zou3519/85/head

Differential Revision: [D16621829](https://our.internmc.facebook.com/intern/diff/D16621829)

zou3519 committed:
Add torch.compiled_with_BUILD_NAMEDTENSOR()

This is a quick check for if pytorch was built with BUILD_NAMEDTENSOR=1.

Test Plan:
- run tests [namedtensor ci]

zou3519 committed:
Update on "Add torch.compiled_with_BUILD_NAMEDTENSOR()"

Add torch.compiled_with_BUILD_NAMEDTENSOR()

This is a quick check for if pytorch was built with BUILD_NAMEDTENSOR=1.

Test Plan:
- run tests [namedtensor ci]

gh-metadata: pytorch pytorch 23623 gh/zou3519/85/head

SsnL commented on torch/__init__.py:
```diff
@@ -334,6 +334,11 @@ def compiled_with_cxx11_abi():
     return _C._GLIBCXX_USE_CXX11_ABI
 
 
+def compiled_with_BUILD_NAMEDTENSOR():
```
does this have to be a public API (i.e., no underscores)?

zou3519 replied:
Nope, I'll make it private, thanks for pointing that out

xuhdev replied:
Should we also let user have a way to know it?

zou3519 replied:
I don't have any good ideas for how to make sure users don't use it aside from adding a docstring. I could just use `torch._C._BUILD_NAMEDTENSOR` directly and hope the double underscore is deterring enough...

xuhdev replied:
Oops, what I meant is that shouldn't we let user have a clean way to know whether the build can has named tensor built in...

zou3519 replied:
> Oops, what I meant is that shouldn't we let user have a clean way to know whether the build can has named tensor built in...

Sorry, I still don't understand the suggestion. The BUILD_NAMEDTENSOR flag is temporary and will be removed sometime in the future. A user should never have to check if they're using a "named tensor build" or a regular pytorch build.

xuhdev replied:
OK, I didn't know that. I vaguely remember that named tensor hurts performance a lot and that's why the build flag is there. If it's only temporary I don't see a problem.

zou3519 replied:
> named tensor hurts performance a lot and that's why the build flag is there.

The build flag is there to prevent any potential performance regressions while I throw the project together.

zou3519 committed:
Update on "Add torch._compiled_with_BUILD_NAMEDTENSOR()"

Add torch._compiled_with_BUILD_NAMEDTENSOR()

This is a quick check for if pytorch was built with BUILD_NAMEDTENSOR=1.

Test Plan:
- run tests [namedtensor ci]

gh-metadata: pytorch pytorch 23623 gh/zou3519/85/head

zou3519 committed:
Update on "Add torch._C._BUILD_NAMEDTENSOR()"

Add torch._C._BUILD_NAMEDTENSOR()

This is a quick, not-user-facing check for if pytorch was built with BUILD_NAMEDTENSOR=1.

Test Plan:
- run tests [namedtensor ci]

gh-metadata: pytorch pytorch 23623 gh/zou3519/85/head
